api.go

```diff
@@ -830,3 +834,105 @@ func (c Client) makeTargetURL(bucketName, objectName, bucketLocation string, que
 
 	return u, nil
 }
+
+// isVirtualHostedStyle - returns true if server supports
+// virtual hosted-style urls
+func (c *Client) isVirtualHostedStyle(bucketName string) bool {
+	if c.endpointURL == sentinelURL {
+		return false
+	}
+	// bucketName can be valid but '.' in the hostname will fail SSL
+	// certificate validation. So do not use host-style for such buckets.
+	if c.endpointURL.Scheme == "https" && strings.Contains(bucketName, ".") {
+		return false
+	}
+	return c.isVirtualHostStyle
+}
+
+// setVirtualHostStyle - auto probes if server supports virtual
+// hosted-style urls
+func (c *Client) setVirtualHostStyle() {
+	// set default to false
+	c.isVirtualHostStyle = false
+
+	host := c.endpointURL.Host
+	scheme := c.endpointURL.Scheme
+	method := "HEAD"
+	// Strip port 80 and 443 so we won't send these ports in Host header.
+	if h, p, err := net.SplitHostPort(host); err == nil {
+		if scheme == "http" && p == "80" || scheme == "https" && p == "443" {
+			host = h
+		}
+	}
+	bucketName := "auto-probe-bucketname"
```

Comment:
What is this `auto-probe-bucketname` ? (I am being lazy but it also means this code needs more comments)

Reply:
this is just a non existent bucket - updated comments to make it self explanatory.